src/module/migration/migrations/882-spell-data-reorganization.ts

```diff
@@ -0,0 +1,270 @@
+import { SaveType } from "@actor/types.ts";
+import { SAVE_TYPES } from "@actor/values.ts";
+import { ItemSourcePF2e } from "@item/base/data/index.ts";
+import { SpellDamageSource, SpellSystemSource } from "@item/spell/data.ts";
+import { MagicTradition, SpellTrait } from "@item/spell/types.ts";
+import { DamageCategoryUnique, DamageType, MaterialDamageEffect } from "@system/damage/types.ts";
+import { isObject, tupleHasValue } from "@util";
+import * as R from "remeda";
+import { MigrationBase } from "../base.ts";
+import { ActorSourcePF2e } from "@actor/data/index.ts";
+
+/** Simplify data structure of spells, add damage `kinds` */
+export class Migration882SpellDataReorganization extends MigrationBase {
+    static override version = 0.882;
+
+    #ensureTraitsPresence(system: MaybeOldSpellSystemSource): { value: SpellTrait[] } {
+        return mergeObject({ value: [] }, system.traits ?? { value: [] });
+    }
+
+    override async updateItem(
+        source: DeepPartial<ItemSourcePF2e>,
+        actorSource?: ActorSourcePF2e,
+        /** Whether this is the top level of the spell item rather than internal partial data */
+        { topLevel = true } = {},
+    ): Promise<void> {
+        if (source.type !== "spell") return;
+        const system: MaybeOldSpellSystemSource = source.system ?? {};
+
+        // Flatten traditions object
+        if (isObject(system.traditions) && "value" in system.traditions && Array.isArray(system.traditions.value)) {
+            system.traits = this.#ensureTraitsPresence(system);
+            system.traits.traditions = [...system.traditions.value].sort();
+        }
+        if ("traditions" in system) system["-=traditions"] = null;
+
+        // Add focus trait if a focus spell
+        if (system.category?.value === "focus") {
```

Comment:
Will need to check that its not a cantrip. This will convert any homebrew focus cantrip that don't have a class name into regular spells, but I don't think that's avoidable. Hopefully its a small number.